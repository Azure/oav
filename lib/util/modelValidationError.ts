// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import { BaseValidationError } from "./baseValidationError"
import { NodeError } from "./validationError"

export interface ModelValidationError
  extends BaseValidationError<ModelValidationError>,
    NodeError<ModelValidationError> {
  operationId?: string
  scenario?: string
  responseCode?: string
  readonly innerErrors?: ModelValidationError[]
}
