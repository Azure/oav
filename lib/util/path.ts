// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import { flatMap } from "@ts-common/iterator"

export const splitPathAndReverse = (p: string | undefined) =>
  p === undefined ? undefined : Array.from(flatMap(p.split("/"), s => s.split("\\"))).reverse()

export const isSubPath = (
  mainPath: ReadonlyArray<string> | undefined,
  subPath: ReadonlyArray<string> | undefined,
) =>
  // return `true` if there are no subPath.
  subPath === undefined ||
  subPath.length === 0 ||
  // return `true` if `subPath` is a sub-path of `mainPath`.
  (mainPath !== undefined &&
    mainPath.length > subPath.length &&
    subPath.every((s, i) => mainPath[i] === s)
  )
