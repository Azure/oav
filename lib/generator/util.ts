import * as fs from "fs";
import * as path from "path";
import { URL } from "url";

const ATTR_MAP: any = {
  location: "location",
  "azure-async-operation": "Azure-AsyncOperation",
  "original-uri": "original-uri",
};

export function getBaseName(filePath: string) {
  return path.basename(filePath);
}
export function isObject(definitionSpec: any) {
  return (
    definitionSpec.type === "object" || "properties" in definitionSpec || "allOf" in definitionSpec
  );
}
export function randomKey() {
  return `key${Math.ceil(Math.random() * 10000)}`;
}

export function getRp(specFilePath: string): string {
  return "microsoft." + specFilePath.split("/").slice(-1)[0].split(".")[0].toLowerCase();
}

export function fileExist(...pathName: string[]): boolean {
  return fs.existsSync(path.resolve(...pathName));
}

export function isLongRunning(specItem: any) {
  return specItem.content["x-ms-long-running-operation"];
}

export function readPayloadFile(payloadDir: string, operationId: string): any {
  if (!fileExist(payloadDir, operationId)) {
    return null;
  }
  const inputPath = path.resolve(payloadDir, operationId);
  const filenames: string[] = fs.readdirSync(inputPath).filter((filename: string) => {
    return (
      filename.endsWith(".json") &&
      ["200", "201", "202", "204"].includes(filename.split(".json")[0])
    );
  });
  if (filenames.length === 0) {
    return null;
  }
  const payload: any = {};
  filenames.forEach((filename: string) => {
    const statusCode = filename.split(".json")[0];
    const data = fs.readFileSync(path.resolve(inputPath, filename));
    payload[statusCode] = JSON.parse(data.toString());
  });
  return payload;
}

export function getPollingAttr(specItem: any) {
  if (!specItem.content["x-ms-long-running-operation"]) {
    return;
  }
  const lrOption = specItem.content["x-ms-long-running-operation-options"];
  if (lrOption && lrOption["final-state-via"]) {
    return ATTR_MAP[lrOption["final-state-via"]];
  }

  return "location";
}

export function getPollingUrl(payload: any, specItem: any) {
  let attr = "location";
  const asyncOperation: string = payload.liveResponse.headers["azure-AsyncOperation"];

  const lrOption = specItem.content["x-ms-long-running-operation-options"];
  let location = payload.liveResponse.headers.location;

  if (lrOption && lrOption["final-state-via"]) {
    attr = ATTR_MAP[lrOption["final-state-via"]];
    location = payload.liveResponse.headers[attr];
  } else if (asyncOperation) {
    location = asyncOperation;
  }
  const url = new URL(location);
  return url.pathname + url.search;
}

export function updateExmAndSpecFile(
  example: any,
  newSpec: any,
  specFilePath: string,
  exampleName: string
): any {
  const outputDir = path.resolve(specFilePath, "../", "examples");
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir);
  }
  const outputPath = path.resolve(outputDir, exampleName);
  console.log("example file path: " + outputPath);
  fs.writeFileSync(outputPath, JSON.stringify(example, null, 2), "utf8");
  if (newSpec) {
    //log.info("updated swagger file path: " + specFilePath);
    fs.writeFileSync(specFilePath, JSON.stringify(newSpec, null, 2), "utf8");
  }
}

export function referenceExmInSpec(
  specFilePath: string,
  apiPath: string,
  methodName: string,
  exampleName: string
): any {
  const data = fs.readFileSync(specFilePath);
  const spec = JSON.parse(data.toString());
  if (!spec.paths[apiPath][methodName]["x-ms-examples"]) {
    spec.paths[apiPath][methodName]["x-ms-examples"] = {};
  }
  if (!(exampleName in spec.paths[apiPath][methodName]["x-ms-examples"])) {
    spec.paths[apiPath][methodName]["x-ms-examples"][exampleName] = {
      $ref: `./examples/${exampleName}.json`,
    };
    return spec;
  }
}
