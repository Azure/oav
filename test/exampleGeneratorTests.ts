import assert from "assert";
import * as fs from "fs";
import * as path from "path";
import ExampleGenerator from "../lib/generator/exampleGenerator";
import { ModelValidationError } from "../lib/util/modelValidationError";
import { generateExamples } from "../lib/validate";
import { log } from "../lib/util/logging";

const payloadDir = `test/exampleGenerator/payloads`;
const specRepoDir = `azure-rest-api-specs`;
jest.setTimeout(999999);
describe.skip("mock examples", () => {
  const specFilePaths = getSpecFilePaths(specRepoDir);
  for (const specPath of specFilePaths) {
    const specInfo = specPath.split("/");
    const apiVersion = specInfo[specInfo.length - 2];
    const rp = specInfo[specInfo.length - 1].split(".")[0];
    it(`spec for ${rp + apiVersion}`, async () => {
      const errors: readonly ModelValidationError[] = await new ExampleGenerator(
        specPath,
        path.resolve(payloadDir, rp + apiVersion)
      ).generateAll();
      if (errors.length > 0) {
        console.error(errors);
      }
      assert.strictEqual(errors.length, 0);
    });
  }
});

describe.skip("test generate example", () => {
  const originalError = log.error;
  //const originalLog = log.info;
  const originalWarn = log.warn;
  let consoleOutput: any[] = [];
  const recordOutput = (output: any) => {
    if (typeof output === "string" && output.indexOf("\\") !== -1) {
      const result = output.replace(/\\/gi, "/");
      consoleOutput.push(result);
    } else {
      consoleOutput.push(output);
    }
  };
  //const mockedLog = (output: any) => recordOutput(output);
  const mockedError = (output: any) => recordOutput(output);
  const mockedWarn = (output: any) => recordOutput(output);
  beforeAll(() => {
    consoleOutput = [];
    //log.info = mockedLog as any
    log.error = mockedError as any;
    log.warn = mockedWarn as any;
  });

  afterAll(() => {
    log.error = originalError;
    //log.info = originalLog;
    log.warn = originalWarn;
  });

  test.each<string[]>([
    ["sql", "package-pure-2020-02-preview"],
    ["signalr", "package-2020-05-01"],
    ["eventgrid", "package-2020-06"],
  ])(
    "from payload,rp:%s",
    async (resourceProviderName, tag) => {
      await generateExamples(
        "",
        payloadDir,
        undefined,
        `test/exampleGenerator/specification/${resourceProviderName}/resource-manager/readme.md`,
        tag
      );
      expect(consoleOutput).toMatchSnapshot(`,tag:${tag}`);
    },
    1000000
  );

  test.each<string[]>([
    ["sql", "package-pure-2020-02-preview"],
    ["signalr", "package-2020-05-01"],
    ["eventgrid", "package-2020-06"],
  ])(
    "from mocker,readme:%s",
    async (resourceProviderName, tag) => {
      await generateExamples(
        "",
        undefined,
        undefined,
        `test/exampleGenerator/specification/${resourceProviderName}/resource-manager/readme.md`,
        tag
      );
      expect(consoleOutput).toMatchSnapshot(`,tag:${tag}`);
    },
    1000000
  );
});

export function getSpecFilePaths(repoDir: string) {
  const rpList = fs.readdirSync(path.resolve(repoDir, "specification"));

  const specPath: string[] = [];
  rpList.forEach((rp: string) => {
    if (!fs.existsSync(path.resolve(repoDir, "specification", rp, "resource-manager"))) {
      return;
    }
    const fullRps = fs
      .readdirSync(path.resolve(repoDir, "specification", rp, "resource-manager"))
      .filter((s) => s.startsWith("Microsoft."));
    if (fullRps.length !== 1) {
      return;
    }
    const stablePath = path.resolve(
      repoDir,
      "specification",
      rp,
      "resource-manager",
      fullRps[0],
      "stable"
    );
    if (!fs.existsSync(stablePath)) {
      return;
    }
    const versions = fs.readdirSync(stablePath).sort((i, j) => j.localeCompare(i));
    if (versions.length === 0) {
      return;
    }
    const filePath = path.resolve(stablePath, versions[0], `${rp}.json`);
    if (!fs.existsSync(filePath)) {
      return;
    }
    specPath.push(filePath);
  });
  return specPath;
}
