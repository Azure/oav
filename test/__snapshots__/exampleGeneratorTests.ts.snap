// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test generate example from mocker,readme:eventgrid: ,tag:package-2020-06 1`] = `
Array [
  Object {
    "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
    "id": "OAV130",
    "message": "Response statusCode \\"201\\" for operation \\"ManagedDatabases_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
  },
  "Found errors in validating the response with statusCode \\"201\\" for x-ms-example \\"ManagedDatabases_CreateOrUpdate_MaximumSet_Gen\\" in operation \\"ManagedDatabases_CreateOrUpdate\\".:
",
  "the validation raised below error:",
  Array [
    Object {
      "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
      "details": Object {
        "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
        "id": "OAV130",
        "message": "Response statusCode \\"201\\" for operation \\"ManagedDatabases_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
      },
      "operationId": "ManagedDatabases_CreateOrUpdate",
      "responseCode": "201",
      "scenario": "ManagedDatabases_CreateOrUpdate_MaximumSet_Gen",
      "severity": 0,
      "source": "response",
    },
  ],
  Object {
    "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
    "id": "OAV130",
    "message": "Response statusCode \\"201\\" for operation \\"ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
  },
  "Found errors in validating the response with statusCode \\"201\\" for x-ms-example \\"ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate_MaximumSet_Gen\\" in operation \\"ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate\\".:
",
  "the validation raised below error:",
  Array [
    Object {
      "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
      "details": Object {
        "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
        "id": "OAV130",
        "message": "Response statusCode \\"201\\" for operation \\"ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
      },
      "operationId": "ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate",
      "responseCode": "201",
      "scenario": "ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate_MaximumSet_Gen",
      "severity": 0,
      "source": "response",
    },
  ],
  Object {
    "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
    "id": "OAV130",
    "message": "Response statusCode \\"201\\" for operation \\"ServerAzureADOnlyAuthentications_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
  },
  "Found errors in validating the response with statusCode \\"201\\" for x-ms-example \\"ServerAzureADOnlyAuthentications_CreateOrUpdate_MaximumSet_Gen\\" in operation \\"ServerAzureADOnlyAuthentications_CreateOrUpdate\\".:
",
  "the validation raised below error:",
  Array [
    Object {
      "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
      "details": Object {
        "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
        "id": "OAV130",
        "message": "Response statusCode \\"201\\" for operation \\"ServerAzureADOnlyAuthentications_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
      },
      "operationId": "ServerAzureADOnlyAuthentications_CreateOrUpdate",
      "responseCode": "201",
      "scenario": "ServerAzureADOnlyAuthentications_CreateOrUpdate_MaximumSet_Gen",
      "severity": 0,
      "source": "response",
    },
  ],
]
`;

exports[`test generate example from mocker,readme:signalr: ,tag:package-2020-05-01 1`] = `
Array [
  Object {
    "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
    "id": "OAV130",
    "message": "Response statusCode \\"201\\" for operation \\"ManagedDatabases_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
  },
  "Found errors in validating the response with statusCode \\"201\\" for x-ms-example \\"ManagedDatabases_CreateOrUpdate_MaximumSet_Gen\\" in operation \\"ManagedDatabases_CreateOrUpdate\\".:
",
  "the validation raised below error:",
  Array [
    Object {
      "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
      "details": Object {
        "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
        "id": "OAV130",
        "message": "Response statusCode \\"201\\" for operation \\"ManagedDatabases_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
      },
      "operationId": "ManagedDatabases_CreateOrUpdate",
      "responseCode": "201",
      "scenario": "ManagedDatabases_CreateOrUpdate_MaximumSet_Gen",
      "severity": 0,
      "source": "response",
    },
  ],
  Object {
    "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
    "id": "OAV130",
    "message": "Response statusCode \\"201\\" for operation \\"ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
  },
  "Found errors in validating the response with statusCode \\"201\\" for x-ms-example \\"ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate_MaximumSet_Gen\\" in operation \\"ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate\\".:
",
  "the validation raised below error:",
  Array [
    Object {
      "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
      "details": Object {
        "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
        "id": "OAV130",
        "message": "Response statusCode \\"201\\" for operation \\"ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
      },
      "operationId": "ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate",
      "responseCode": "201",
      "scenario": "ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate_MaximumSet_Gen",
      "severity": 0,
      "source": "response",
    },
  ],
  Object {
    "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
    "id": "OAV130",
    "message": "Response statusCode \\"201\\" for operation \\"ServerAzureADOnlyAuthentications_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
  },
  "Found errors in validating the response with statusCode \\"201\\" for x-ms-example \\"ServerAzureADOnlyAuthentications_CreateOrUpdate_MaximumSet_Gen\\" in operation \\"ServerAzureADOnlyAuthentications_CreateOrUpdate\\".:
",
  "the validation raised below error:",
  Array [
    Object {
      "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
      "details": Object {
        "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
        "id": "OAV130",
        "message": "Response statusCode \\"201\\" for operation \\"ServerAzureADOnlyAuthentications_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
      },
      "operationId": "ServerAzureADOnlyAuthentications_CreateOrUpdate",
      "responseCode": "201",
      "scenario": "ServerAzureADOnlyAuthentications_CreateOrUpdate_MaximumSet_Gen",
      "severity": 0,
      "source": "response",
    },
  ],
]
`;

exports[`test generate example from mocker,readme:sql: ,tag:package-pure-2020-02-preview 1`] = `
Array [
  Object {
    "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
    "id": "OAV130",
    "message": "Response statusCode \\"201\\" for operation \\"ManagedDatabases_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
  },
  "Found errors in validating the response with statusCode \\"201\\" for x-ms-example \\"ManagedDatabases_CreateOrUpdate_MaximumSet_Gen\\" in operation \\"ManagedDatabases_CreateOrUpdate\\".:
",
  "the validation raised below error:",
  Array [
    Object {
      "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
      "details": Object {
        "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
        "id": "OAV130",
        "message": "Response statusCode \\"201\\" for operation \\"ManagedDatabases_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
      },
      "operationId": "ManagedDatabases_CreateOrUpdate",
      "responseCode": "201",
      "scenario": "ManagedDatabases_CreateOrUpdate_MaximumSet_Gen",
      "severity": 0,
      "source": "response",
    },
  ],
  Object {
    "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
    "id": "OAV130",
    "message": "Response statusCode \\"201\\" for operation \\"ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
  },
  "Found errors in validating the response with statusCode \\"201\\" for x-ms-example \\"ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate_MaximumSet_Gen\\" in operation \\"ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate\\".:
",
  "the validation raised below error:",
  Array [
    Object {
      "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
      "details": Object {
        "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
        "id": "OAV130",
        "message": "Response statusCode \\"201\\" for operation \\"ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
      },
      "operationId": "ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate",
      "responseCode": "201",
      "scenario": "ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate_MaximumSet_Gen",
      "severity": 0,
      "source": "response",
    },
  ],
  Object {
    "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
    "id": "OAV130",
    "message": "Response statusCode \\"201\\" for operation \\"ServerAzureADOnlyAuthentications_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
  },
  "Found errors in validating the response with statusCode \\"201\\" for x-ms-example \\"ServerAzureADOnlyAuthentications_CreateOrUpdate_MaximumSet_Gen\\" in operation \\"ServerAzureADOnlyAuthentications_CreateOrUpdate\\".:
",
  "the validation raised below error:",
  Array [
    Object {
      "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
      "details": Object {
        "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
        "id": "OAV130",
        "message": "Response statusCode \\"201\\" for operation \\"ServerAzureADOnlyAuthentications_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
      },
      "operationId": "ServerAzureADOnlyAuthentications_CreateOrUpdate",
      "responseCode": "201",
      "scenario": "ServerAzureADOnlyAuthentications_CreateOrUpdate_MaximumSet_Gen",
      "severity": 0,
      "source": "response",
    },
  ],
]
`;

exports[`test generate example from payload,rp:eventgrid: ,tag:package-2020-06 1`] = `
Array [
  Object {
    "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
    "id": "OAV130",
    "message": "Response statusCode \\"201\\" for operation \\"ManagedDatabases_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
  },
  "Found errors in validating the response with statusCode \\"201\\" for x-ms-example \\"ManagedDatabases_CreateOrUpdate_MaximumSet_Gen\\" in operation \\"ManagedDatabases_CreateOrUpdate\\".:
",
  "the validation raised below error:",
  Array [
    Object {
      "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
      "details": Object {
        "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
        "id": "OAV130",
        "message": "Response statusCode \\"201\\" for operation \\"ManagedDatabases_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
      },
      "operationId": "ManagedDatabases_CreateOrUpdate",
      "responseCode": "201",
      "scenario": "ManagedDatabases_CreateOrUpdate_MaximumSet_Gen",
      "severity": 0,
      "source": "response",
    },
  ],
  Object {
    "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
    "id": "OAV130",
    "message": "Response statusCode \\"201\\" for operation \\"ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
  },
  "Found errors in validating the response with statusCode \\"201\\" for x-ms-example \\"ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate_MaximumSet_Gen\\" in operation \\"ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate\\".:
",
  "the validation raised below error:",
  Array [
    Object {
      "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
      "details": Object {
        "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
        "id": "OAV130",
        "message": "Response statusCode \\"201\\" for operation \\"ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
      },
      "operationId": "ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate",
      "responseCode": "201",
      "scenario": "ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate_MaximumSet_Gen",
      "severity": 0,
      "source": "response",
    },
  ],
  Object {
    "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
    "id": "OAV130",
    "message": "Response statusCode \\"201\\" for operation \\"ServerAzureADOnlyAuthentications_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
  },
  "Found errors in validating the response with statusCode \\"201\\" for x-ms-example \\"ServerAzureADOnlyAuthentications_CreateOrUpdate_MaximumSet_Gen\\" in operation \\"ServerAzureADOnlyAuthentications_CreateOrUpdate\\".:
",
  "the validation raised below error:",
  Array [
    Object {
      "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
      "details": Object {
        "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
        "id": "OAV130",
        "message": "Response statusCode \\"201\\" for operation \\"ServerAzureADOnlyAuthentications_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
      },
      "operationId": "ServerAzureADOnlyAuthentications_CreateOrUpdate",
      "responseCode": "201",
      "scenario": "ServerAzureADOnlyAuthentications_CreateOrUpdate_MaximumSet_Gen",
      "severity": 0,
      "source": "response",
    },
  ],
]
`;

exports[`test generate example from payload,rp:signalr: ,tag:package-2020-05-01 1`] = `
Array [
  Object {
    "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
    "id": "OAV130",
    "message": "Response statusCode \\"201\\" for operation \\"ManagedDatabases_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
  },
  "Found errors in validating the response with statusCode \\"201\\" for x-ms-example \\"ManagedDatabases_CreateOrUpdate_MaximumSet_Gen\\" in operation \\"ManagedDatabases_CreateOrUpdate\\".:
",
  "the validation raised below error:",
  Array [
    Object {
      "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
      "details": Object {
        "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
        "id": "OAV130",
        "message": "Response statusCode \\"201\\" for operation \\"ManagedDatabases_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
      },
      "operationId": "ManagedDatabases_CreateOrUpdate",
      "responseCode": "201",
      "scenario": "ManagedDatabases_CreateOrUpdate_MaximumSet_Gen",
      "severity": 0,
      "source": "response",
    },
  ],
  Object {
    "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
    "id": "OAV130",
    "message": "Response statusCode \\"201\\" for operation \\"ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
  },
  "Found errors in validating the response with statusCode \\"201\\" for x-ms-example \\"ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate_MaximumSet_Gen\\" in operation \\"ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate\\".:
",
  "the validation raised below error:",
  Array [
    Object {
      "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
      "details": Object {
        "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
        "id": "OAV130",
        "message": "Response statusCode \\"201\\" for operation \\"ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
      },
      "operationId": "ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate",
      "responseCode": "201",
      "scenario": "ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate_MaximumSet_Gen",
      "severity": 0,
      "source": "response",
    },
  ],
  Object {
    "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
    "id": "OAV130",
    "message": "Response statusCode \\"201\\" for operation \\"ServerAzureADOnlyAuthentications_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
  },
  "Found errors in validating the response with statusCode \\"201\\" for x-ms-example \\"ServerAzureADOnlyAuthentications_CreateOrUpdate_MaximumSet_Gen\\" in operation \\"ServerAzureADOnlyAuthentications_CreateOrUpdate\\".:
",
  "the validation raised below error:",
  Array [
    Object {
      "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
      "details": Object {
        "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
        "id": "OAV130",
        "message": "Response statusCode \\"201\\" for operation \\"ServerAzureADOnlyAuthentications_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
      },
      "operationId": "ServerAzureADOnlyAuthentications_CreateOrUpdate",
      "responseCode": "201",
      "scenario": "ServerAzureADOnlyAuthentications_CreateOrUpdate_MaximumSet_Gen",
      "severity": 0,
      "source": "response",
    },
  ],
]
`;

exports[`test generate example from payload,rp:sql: ,tag:package-pure-2020-02-preview 1`] = `
Array [
  Object {
    "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
    "id": "OAV130",
    "message": "Response statusCode \\"201\\" for operation \\"ManagedDatabases_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
  },
  "Found errors in validating the response with statusCode \\"201\\" for x-ms-example \\"ManagedDatabases_CreateOrUpdate_MaximumSet_Gen\\" in operation \\"ManagedDatabases_CreateOrUpdate\\".:
",
  "the validation raised below error:",
  Array [
    Object {
      "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
      "details": Object {
        "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
        "id": "OAV130",
        "message": "Response statusCode \\"201\\" for operation \\"ManagedDatabases_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
      },
      "operationId": "ManagedDatabases_CreateOrUpdate",
      "responseCode": "201",
      "scenario": "ManagedDatabases_CreateOrUpdate_MaximumSet_Gen",
      "severity": 0,
      "source": "response",
    },
  ],
  Object {
    "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
    "id": "OAV130",
    "message": "Response statusCode \\"201\\" for operation \\"ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
  },
  "Found errors in validating the response with statusCode \\"201\\" for x-ms-example \\"ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate_MaximumSet_Gen\\" in operation \\"ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate\\".:
",
  "the validation raised below error:",
  Array [
    Object {
      "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
      "details": Object {
        "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
        "id": "OAV130",
        "message": "Response statusCode \\"201\\" for operation \\"ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
      },
      "operationId": "ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate",
      "responseCode": "201",
      "scenario": "ManagedInstanceAzureADOnlyAuthentications_CreateOrUpdate_MaximumSet_Gen",
      "severity": 0,
      "source": "response",
    },
  ],
  Object {
    "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
    "id": "OAV130",
    "message": "Response statusCode \\"201\\" for operation \\"ServerAzureADOnlyAuthentications_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
  },
  "Found errors in validating the response with statusCode \\"201\\" for x-ms-example \\"ServerAzureADOnlyAuthentications_CreateOrUpdate_MaximumSet_Gen\\" in operation \\"ServerAzureADOnlyAuthentications_CreateOrUpdate\\".:
",
  "the validation raised below error:",
  Array [
    Object {
      "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
      "details": Object {
        "code": "RESPONSE_BODY_NOT_IN_EXAMPLE",
        "id": "OAV130",
        "message": "Response statusCode \\"201\\" for operation \\"ServerAzureADOnlyAuthentications_CreateOrUpdate\\" has no response body provided in the example, however the response does have a \\"schema\\" defined in the swagger spec.",
      },
      "operationId": "ServerAzureADOnlyAuthentications_CreateOrUpdate",
      "responseCode": "201",
      "scenario": "ServerAzureADOnlyAuthentications_CreateOrUpdate_MaximumSet_Gen",
      "severity": 0,
      "source": "response",
    },
  ],
]
`;
